packages/service-core/src/sync/sync.ts

```diff
@@ -103,138 +108,222 @@ async function* streamResponseInner(
+          ...e,
+          priority: dataBuckets.get(e.bucket)!.description!.priority
+        }));
+        bucketsToFetch = updatedBucketDescriptions;
+
+        let message = `Updated checkpoint: ${checkpoint} | `;
+        message += `write: ${writeCheckpoint} | `;
+        message += `buckets: ${allBuckets.length} | `;
+        message += `updated: ${limitedBuckets(diff.updatedBuckets, 20)} | `;
+        message += `removed: ${limitedBuckets(diff.removedBuckets, 20)}`;
+        logger.info(message, {
+          checkpoint,
+          user_id: syncParams.user_id,
+          buckets: allBuckets.length,
+          updated: diff.updatedBuckets.length,
+          removed: diff.removedBuckets.length
+        });
+
+        const checksum_line: util.StreamingSyncCheckpointDiff = {
+          checkpoint_diff: {
+            last_op_id: checkpoint,
+            write_checkpoint: writeCheckpoint ? String(writeCheckpoint) : undefined,
+            removed_buckets: diff.removedBuckets,
+            updated_buckets: updatedBucketDescriptions
+          }
+        };
 
-      yield checksum_line;
-    } else {
-      let message = `New checkpoint: ${checkpoint} | write: ${writeCheckpoint} | `;
-      message += `buckets: ${allBuckets.length} ${limitedBuckets(allBuckets, 20)}`;
-      logger.info(message, { checkpoint, user_id: syncParams.user_id, buckets: allBuckets.length });
-      bucketsToFetch = allBuckets;
-      const checksum_line: util.StreamingSyncCheckpoint = {
-        checkpoint: {
-          last_op_id: checkpoint,
-          write_checkpoint: writeCheckpoint ? String(writeCheckpoint) : undefined,
-          buckets: [...checksumMap.values()]
+        yield checksum_line;
+      } else {
+        let message = `New checkpoint: ${checkpoint} | write: ${writeCheckpoint} | `;
+        message += `buckets: ${allBuckets.length} ${limitedBuckets(allBuckets, 20)}`;
+        logger.info(message, { checkpoint, user_id: syncParams.user_id, buckets: allBuckets.length });
+        bucketsToFetch = allBuckets;
+        const checksum_line: util.StreamingSyncCheckpoint = {
+          checkpoint: {
+            last_op_id: checkpoint,
+            write_checkpoint: writeCheckpoint ? String(writeCheckpoint) : undefined,
+            buckets: [...checksumMap.values()].map((e) => ({
+              ...e,
+              priority: dataBuckets.get(e.bucket)!.description!.priority
+            }))
+          }
+        };
+        yield checksum_line;
+      }
+      lastChecksums = checksumMap;
+      lastWriteCheckpoint = writeCheckpoint;
+
+      // Start syncing data for buckets up to the checkpoint. As soon as we have completed at least one priority and
+      // at least 1000 operations, we also start listening for new checkpoints concurrently. When a new checkpoint comes
+      // in while we're still busy syncing data for lower priorities, interrupt the current operation and start syncing
+      // the new checkpoint.
```

Comment:
> As soon as we have completed at least one priority

Wouldn't that negate most of the advantages here? I.e. if we're busy syncing bulk data with a low priority, that may never be interrupted for a new checkpoint with higher-priority buckets.

Reply:
That's a good point, I think we should just check for 1000 rows synced then.

My original idea (which this check doesn't implement accurately) was that we shouldn't interrupt high-priority sync operations when we receive a new checkpoint that might only cover lower-priority buckets.
I'm not sure if we should still check this separately now or if it doesn't matter in practice because we expect high-priority buckets to be generally smaller (and would still have the 1000 row limit before we listen for a new checkpoint).